install.sh

```diff
@@ -575,19 +645,131 @@ get_linux_distro() {
   fi
+  echo "$distro"
+}
+
+install_package() {
+  local package="$1"
+  distro=$(get_linux_distro)
+  
+  log::info "Detected Linux distribution: $distro"
+
+  case "$distro" in
+    *CentOS*|*Red\ Hat*|*Fedora*)
+      run_as_root "yum -y install $package"
+      ;;
+    *Debian*|*Ubuntu*)
+      run_as_root "apt-get -y update"
+      run_as_root "DEBIAN_FRONTEND=noninteractive apt-get -y install $package"
+      ;;
+    *SUSE*|*openSUSE*)
+      run_as_root "zypper install $package"
+      ;;
+    *RHEL*)
+      if [[ "$distro" == *"7"* ]]; then
+        run_as_root "yum -y install $package"
+      elif [[ "$distro" == *"8"* ]]; then
+        run_as_root "dnf module install -y $package"  
+      fi
+      ;;
+    *)
+      abort "Unsupported distribution: $distro"
+      ;;
+  esac
+}
+
+install_buildah(){
+  distro=$(get_linux_distro)
+  log::info "Installing Buildah and configuring user namespaces..."
+
+  case "$distro" in
+    *Debian*|*Ubuntu*)
+      install_package "buildah uidmap"
+      ;;
+    *CentOS*|*Red\ Hat*|*Fedora*)
+      run_as_root "yum groupinstall -y 'Development Tools'"
+      install_package "curl-devel expat-devel gettext-devel openssl-devel zlib-devel perl autoconf"
+      ;;
+    *SUSE*|*openSUSE*)
+      install_package "buildah shadow"
+      ;;
+    *RHEL*)
+      if [[ "$distro" == *"8"* ]]; then
+          run_as_root "dnf module enable -y container-tools:2.0"
+          install_package "buildah uidmap"
+      elif [[ "$distro" == *"7"* ]]; then
+          run_as_root "subscription-manager repos --enable=rhel-7-server-extras-rpms"
+          install_package "buildah uidmap"
+      fi
+      ;;
+    *)
+      abort "Unsupported distribution."
+      ;;
+  esac
+
+  set_user_subuids_subgids
+
+  # Migrate Podman system if applicable
+  if is_command_exists podman; then
+    log::info "Migrating Podman system configuration..."
+    run_as_root "podman system migrate"
+  fi
+
+  log::info "Buildah installation and configuration completed."
+}
+
+get_free_range() {
+  
+  local start=100000
+  local end=600000
+  local range_size=65536
+
+  local used_ranges
+  used_ranges=$(awk -F: '{print $2 ":" $3}' /etc/subuid /etc/subgid | sort -n)
+
+  while read -r range; do
+      local range_start range_end
+      range_start=$(echo "$range" | cut -d: -f1)
+      range_end=$((range_start + $(echo "$range" | cut -d: -f2) - 1))
+
+      if ((start + range_size - 1 < range_start)); then
+          echo "$start:$range_size"
+          return 0
+      fi
+      start=$((range_end + 1))
+  done <<< "$used_ranges"
+
+  echo "$start:$range_size"
+  return 0
+}
+
+set_user_subuids_subgids() {
 
-  printf '%s' "$distro" | tr '[:upper:]' '[:lower:]'
+  if grep -q "^$(get_user):" /etc/subuid; then
+    log::info "User $(get_user) already has a subuid range."
+    return 0
+  fi
+
+  if grep -q "^$(get_user):" /etc/subgid; then
+    log::info "User $(get_user) already has a subgid range."
+    return 0
+  fi
```

Comment:
Also, what if the range is here for the current user, but it's not big enough? Let's say 1k instead of 65k. Let's abort with an error if this is the case.